### ISSUE TITLE
Delete Fixed-Form source format

### ISSUE THREAD

**Issue by jacobwilliams** [j3-fortran/fortran_proposals#108]
It's long past time. Like implicit typing, fixed-form source is a massive embarrassment to modern Fortran that needs to go.

Provide a canonical fixed to free conversion script (see also Python's [2to3](https://docs.python.org/2/library/2to3.html)).

**Comment by certik** [comment]
I agree. It might be that compilers will support it for a long time for practical reasons, but the standard should simply say that fixed-form is deprecated and provide a canonical conversion tool that works with any code. That way the community will simply move code over (automatically).

However I would not use Python 2 to 3 as an example, as that script didn't actually work properly, due to the dynamic nature of Python and caused projects to invest insane amount of efforts to do the port. See my blog post with details: https://ondrejcertik.blogspot.com/2013/08/how-to-support-both-python-2-and-3.html. On the other hand, for the fixed-form, I think there might be a way to do this fully automatically and robustly --- but we need to do try this and ensure it works with any code out there, before deprecating fixed-form.

**Comment by FortranFan** [comment]
@certik wrote:

>.. the standard should simply say that fixed-form is deprecated ..

The standard currently marks fixed-form source as **obscolescent**.  There is a section in the standard for **deleted**, but there is no deprecated.  Don't know if any ISO or INCITS considerations come into play here when it comes to these kinds of classification.

There appears an extremely strong resistance to mark features as deleted.

But the advantages of deletions are obvious: once something is deleted, the standard text (e.g., section 6 in current standard) no longer needs to cover that feature plus a compiler implementation, especially a new one, can decide not to support those deleted features.  This might free up development resources to focus on new features while attracting certain customers (I know of some) who want nothing whatsoever to do with legacy aspects and want their code as well as compiler tool chains to focus exclusively on modern coding practices.

**Comment by jacobwilliams** [comment]
Delete it!

**Comment by klausler** [comment]
Production Fortran compilers don't have the luxury to delete features that paying customers are using, standard or not.  Fixed form isn't hard to implement, it's still in use, and it's not going to leave the actual Fortran language, so removing it from the standard document just makes its specification harder to find in the future.

**Comment by milancurcic** [comment]
@jacobwilliams Proposals like this must have technical arguments. What's the problem with fixed form and what would be the benefits of removing it from the language? What would be the costs?

**Comment by gronki** [comment]
Well I think stupid line restrictions that make free form not actually free
are more of a concern.

Dominik

wt., 10 gru 2019, 20:03 użytkownik Milan Curcic <notifications@github.com>
napisał:

**Comment by FortranFan** [comment]
@gronki  wrote:

> Well I think stupid line restrictions that make free form not actually free are more of a concern. ..

Not sure what the above statement refers to.

With respect to Fortran 202X the next standard revision, thankfully at the Tokyo meeting, WG5 has approved the following specifications for free-form source:

```
3 Specifications

A source line contains at most ten thousand characters.

A source statement contains at most one million characters.

Processors are required to be able to report too-long lines.

Continuation is unchanged, except that the limit of 255 is removed.
```

Barring any surprises during any further development of Fortran 202X work-list by J3 or reconsideration during subsequent WG5 meetings which, I think, would be unlikely, Fortran 202X should considerably "free up" the source form.

Fortran 202X development with US01 worklist item on "Allow much longer statement lines and overall statement length" is a real quick grab of low-hanging fruit of a language development that should alleviate concerns of actual coders as well as any issue that might arise with machine-generated code.  

Readers can thus view free-form source as a mostly (if they aren't inclined to accept as fully) solved problem) and focus back on the original post with fixed-form source

**Comment by rweed** [comment]
Doing the most basic conversions required to transform a fixed source file to a free format one (ie changing comment cards to ! and continuations) is actually fairly trivial. I wrote my own code (in Fortran) to do that 20 years ago. Most current Fortran books suggest a strategy for continuations of placing an ampersand in column 73 on the continued line and ampersands in column 6 for the continuation. Detecting continuations is also trivial (just look for characters in column 6). The hard part of conversion though is dealing with what I call Stupid Programmer Tricks (ala Lettermans Stupid Human Tricks) where folks insisted on inserting white spaces between characters in reserved words etc. My "favorite" one was in a very large DoE program where one of the programmers (I think a grad student or a post-doc who didn't know better) thought it was jolly fun to  put spaces between the letters of subroutine and the subroutine name ie
S U B R O U T I N E   M A T M O D. The real complexity comes when you want to shift things over to column one and then refactor the code to some indentation pattern that suits your programming style. Also if you want to lower case on all or some of the text to get away from the ALL UPPER CASE shouting, you have to deal with detecting and handleing character strings where you might not want the case to change.  Still all of this can be handled by a competent Python/C++/Java (and even Fortran) programmer or any one (on Linux) with sed/awk and bash scripting skills. Going up to the next level and generating a tool to do a complete "makeover" that will convert DO loops to block constructs, change GOTOs to If/Then/Else is a much harder task and probably requires using the current compiler front ends and the ASTs etc generated as the first pass. Maybe some tools to do this can be written for LFortran and flang/f18 using LLVM but I know very little about LLVM so I'll let the experts comment on that.

**Comment by ivan-pi** [comment]
Don't any of you have those days where you feel like taking an old 1966 dusty deck in fixed-form source format out of the closet and having it compile smoothly with the current Fortran compilers? ;) It is what I do occasionally on weekends (like for example here: https://github.com/ivan-pi/stroud_quad).

I agree with @milancurcic that we need some technical arguments. I don't see fixed-form as an embarrassment and most people learning Fortran start using free-form directly (or at least I hope so!).

I can perhaps see a technical argument, that currently compilers assume .f files are in fixed-form, leading to the problem that @sblionel blogged about: https://software.intel.com/en-us/blogs/2013/01/11/doctor-fortran-in-source-form-just-wants-to-be-free. Maybe this discussion should be about standardizing the file extension instead, like @jacobwilliams has written about previously (see here http://degenerateconic.com/fortran-file-extensions/ ).

If fixed-form was removed/deprecated then we, modern free-form Fortran people, could again use the .f file ending, instead of defaulting to .f90 (which we use just out of convenience since it directly works on both the Intel and gfortran compiler, without the need for extra compiler flags).

**Comment by sblionel** [comment]
When we use the term “deprecated”, this is what the standard calls “obsolescent”, marking it as a candidate for deletion. Deleting a feature doesn’t magically free up something for redefinition, and NO compiler is going to remove support for fixed-form even if it gets deleted.

I would be delighted to find a way out of the mess we got ourselves into in 1991, but easy compatibility with decades-old sources is a major strength of the language, and I would not support anything that hurt that.

**Comment by jacobwilliams** [comment]
The reason is not entirely technical (although there are some technical reasons). Sure, for some people it's great that they can still compile their 50 year old code without modifications, but I think in the long run, it's very bad for the future of the language.

People are not choosing Fortran anymore. Young people are not learning it. Most people still equate Fortran with FORTRAN 77. One of the reasons for this is the vast amount of FORTRAN 77 code out there in the wild. Look at all the old FORTRAN 77 code that has been wrapped via SciPy. People see this stuff and think that's what Fortran is (and they know that they want nothing to do with it). It would be better if this old stuff just wasn't compilable anymore. Then people would either just abandon it (which is going to happen anyway at some point), or modernize it.

Backwards compatibility is great, but at some point it's time to move on. The time has long passed. People complain about the Python 2-3 transition, but in the end everyone is going to just deal with it. Python 3 probably has more users than all of Fortran, and conversion from fixed to free form is not nearly as difficult as converting from Python 2 to 3 (not even close). That's why an official conversion routine would be vital.

I know the committee doesn't agree, but getting new people onboard should be more important than backward compatibility. I see fixed-form source (along with implict save and other atavistic nonsense) as a gangrenous limb that needs to be removed before it kills the patient. :)

**Comment by klausler** [comment]
> The reason is not entirely technical (although there are some technical reasons). 

What are these technical reasons?

**Comment by jacobwilliams** [comment]
It seems like I've seen some discussions here about new features and worrying that some things aren't possible because of some limitation of fixed-form source?

**Comment by gronki** [comment]
Yes I definely remember such arguments too although I can't bring up
particular features or links. But it was mostly regarding syntax
simplification related topics (along the lines of proposal X can't be done
because fixed form does not see spaces blah blah blah). Maybe someone can
help out here.

śr., 11 gru 2019, 20:39 użytkownik Jacob Williams <notifications@github.com>
napisał:

**Comment by gronki** [comment]
I'm all for killing old code. If Fortran might die too its a risk worth
taking. It is dead anyway and we are just trying to revive it.

śr., 11 gru 2019, 20:53 użytkownik Dominik Gronkiewicz <gronki@gmail.com>
napisał:

**Comment by FortranFan** [comment]
@jacobwilliams wrote:

> It seems like I've seen some discussions here about new features and worrying that some things aren't possible because of some limitation of fixed-form source? Somebody correct me if I'm wrong.

Glad this point has finally surfaced and it's OP who has brought it up!  I was waiting for it and I had deliberately kept quiet on this front in my first response upthread since I expected that to  bring out all the usual non-technical counterarguments to pop up as they have.  Which itself is most ironic given the premise of wanting technical arguments for a request to move forward.

From what I see given how the Fortran standard and its bearers understand matters (perhaps using ISO/INCITS considerations, as to whether that is pursued rightly or wrongly is a separate but legitimate debate), one is left with a **binary** proposition: no deletion of fixed-form source essentially implies continued support of it.  And effectively then, every decision of all new features in the language while fixed-form source continues to be NOT DELETED, particularly at the customer-facing level meaning the syntax which is what the persevering practitioners of Fortran dabble with, is impacted and even held hostage by the peculiarities of that fixed source form.

Consider the current work by J3 on enumerations and see this paper: https://j3-fortran.org/doc/year/19/19-249.txt:
*  Now, there were some good suggestions during the last meeting in Las Vegas on the requirements and specifications of this new feature and how to also allow **compact** declarations of an enumeration type along with its enumerator members e.g., for simple cases, something supported by many popular languages in scientific and technical domains such as Java, etc..
* Unfortunately such suggestions could not be considered and among the reasons for that was the language rules of fixed-form source.
* It appears highly likely what J3 will develop and what WG5 will approve for Fortran 202X toward enumerations is a most limited Pascal v1980 style facility (e.g., no ability to specify the value of enumerators UNLIKE all the top languages per 2019 IEEE survey) BUT with considerable VERBOSITY to boot - I'll be laughed off my domain if I show the ensuing enumeration facility in Fortran around year 2030 or whenever the compiler support for it finally arrives.
* The pernicious impact of fixed-form source would have played no small role in this abomination.

I think OP has legitimate merit with the request here.

**Comment by klausler** [comment]
> @jacobwilliams wrote:
> 
> > It seems like I've seen some discussions here about new features and worrying that some things aren't possible because of some limitation of fixed-form source? Somebody correct me if I'm wrong.
> 
> Glad this point has finally surfaced and it's OP who has brought it up! I was waiting for it and I had deliberately kept quiet on this front in my first response upthread since I expected that to bring out all the usual non-technical counterarguments to pop up as they have. Which itself is most ironic given the premise of wanting technical arguments for a request to move forward.
> 
> From what I see given how the Fortran standard and its bearers understand matters (perhaps using ISO/INCITS considerations, as to whether that is pursued rightly or wrongly is a separate but legitimate debate), one is left with a **binary** proposition: no deletion of fixed-form source essentially implies continued support of it. And effectively then, every decision of all new features in the language while fixed-form source continues to be NOT DELETED, particularly at the customer-facing level meaning the syntax which is what the persevering practitioners of Fortran dabble with, is impacted and even held hostage by the peculiarities of that fixed source form.
> 
> Consider the current work by J3 on enumerations and see this paper: https://j3-fortran.org/doc/year/19/19-249.txt:
> 
> * Now, there were some good suggestions during the last meeting in Las Vegas on the requirements and specifications of this new feature and how to also allow **compact** declarations of an enumeration type along with its enumerator members e.g., for simple cases, something supported by many popular languages in scientific and technical domains such as Java, etc..
> * Unfortunately such suggestions could not be considered and among the reasons for that was the language rules of fixed-form source.
> * It appears highly likely what J3 will develop and what WG5 will approve for Fortran 202X toward enumerations is a most limited Pascal v1980 style facility (e.g., no ability to specify the value of enumerators UNLIKE all the top languages per 2019 IEEE survey) BUT with considerable VERBOSITY to boot - I'll be laughed off my domain if I show the ensuing enumeration facility in Fortran around year 2030 or whenever the compiler support for it finally arrives.
> * The pernicious impact of fixed-form source would have played no small role in this abomination.
> 
> I think OP has legitimate merit with the request here.

If one actually goes and reads the linked paper, one will see that fixed-form is used as a reason for requiring the keyword `ENUMERATOR` to appear before a list of enumerator names; in the absence of the keyword, an sole enumerator name might be misinterpreted as another keyword statement.  Retaining the keyword seems like a good idea for free-form source too for error recovery reasons, so I don't think that fixed-form is exactly constituting a major barrier to technical progress in this case.

And you need to dramatically reduce the hyperbolic gibbering if you want this Fortran implementer to spend more time perusing this site.  The good ideas are getting buried.

**Comment by certik** [comment]
@FortranFan's main point is valid, that fixed form was used as an argument why something cannot be done in 19/19-249.txt. If fixed form didn't have to be considered, perhaps the `ENUMERATOR` keyword would still be necessary for other reasons, as @klausler indicated, but perhaps not. However let's discuss that at #11.

@FortranFan please try hard to stick on topic and just discuss technical arguments in issues, otherwise every issue becomes buried and people stop participating here. If you want to discuss your wider points, just open a new issue and link it.

**Comment by klausler** [comment]
Some real-world data on the cost of supporting fixed-form source in a compiler:

The parsing phase of the f18 compiler comprises 21373 lines of C++ code today; this phase recognizes Fortran programs and represents them in a parse tree data structure, which is constructed in memory prior to the following phase of semantic analysis.  Fixed-form source affects only the part of the parsing code that preprocesses and normalizes the source in memory (stripping comments, joining continued lines, &c.) prior to running the recursive descent backtracking parser proper.

I searched through this code to count the lines that I could delete if fixed form source did not exist, and came up with a total of 250 lines, many of which were comments or just `}`.  In the parser proper, fixed form affects two lines of source (they suppress a warning about missing white space when free form requires it).  And there's roughly 20 lines of code elsewhere that manages the setting and clearing of the flags that indicates fixed-form source.

So that adds up to about 1.3% of the parser, which itself constitutes about 1/3rd of the f18 front-end.

**Comment by sblionel** [comment]
I cannot think of any features proposed that were rejected because of fixed-form's existence. (I don't count syntax that was revised for fixed-form compatibility.)  Keep in mind that every compiler implementor I have spoken with on the topic of deleted features has said that they must continue to support deleted features. Deleting features provides no benefits and actually adds complexity for implementations which have to figure out how those features continue to work with the rest of the standard.

I do not agree that fixed-form source is why (allegedly) young programmers are not learning Fortran. I also disagree with that basic premise. While Fortran is not often taught in the US, in favor of sparkly languages not common in production environments, it is widely taught in Europe and Asia. This isn't the standard's fault.

My recommendation would be to "save your strength" and advocate for new features that will support the programming paradigms you'd like to use.

**Comment by FortranFan** [comment]
@certik wrote:

> ..
> @FortranFan please try hard to stick on topic and just discuss technical arguments in issues, ..

May I please see some consistent adherence and application of this?  The issue I notice across many threads here and it's one of non-technical counter-arguments to original posts.  A basic fact with Fortran, nearly 3 decades after Fortran 90 revision, is that when it comes to certain legacy aspects such as implicit typing, fixed-form source, etc., the requests and suggestions to move forward as in the original post here and which allow the language to progress more freely are **overarching considerations** and can scarcely be put forth using any technical points given how the support toward the legacy aspects permeates across the standard, its development, and implementations in many subtle, hidden, and now 41 years after ANSI document on FORTRAN 77 in overwhelmingly harmful ways.

Under the circumstances, it's those making counterarguments who should have to strive to be technical, particularly from any of the points listed toward "creating a positive environment" under https://github.com/j3-fortran/fortran_proposals/blob/master/CODE_OF_CONDUCT.md.  Otherwise, 

I fail to see anything technical in the arguments posted counter to the original post here: https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564172206, especially https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564182953. https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564370588

So if the matter of some guidelines for counterarguments can be addressed better under CoC of this site and if some consideration can be extended to what I suggest here, it will help.

**Comment by milancurcic** [comment]
Non-technical arguments are OK, and I don't think they should be dismissed or discouraged (or dictated by the CoC). However, technical arguments go a long way to help a proposal. This isn't a competition in which the strongest technical argument wins and gets adopted in the standard. This is about trying to **persuade** the rest of the community (and hopefully committee members) that a change is needed. The original post was a one-liner with almost nothing to give, this is why I asked for some technical arguments so we can have a real discussion and weigh the pros and cons.

When each of us propose a change to the language, we need to do the homework and show (or convincingly speculate) that the benefits will outweigh the cost, not the other way around.

**Comment by klausler** [comment]
> I fail to see anything technical in the arguments posted counter to the original post here: [#108 (comment)](https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564172206), especially [#108 (comment)](https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564182953).

Statements of fact qualify as technical points, whether they're required to be so to constitute counterarguments to non-technical complaints or not.  Specifically, from my comment to which you linked above: fixed-form source is still in use and is easy to support.  The first claim is obviously true, and I've documented why the second claim holds with data.

**Comment by aradi** [comment]
I'd like to note that a language version keyword as discussed in #83 could also allow for giving up compatibility with fixed-form source. Whenever the programmer specifies a language version newer than a given date (e.g. 202Y), the processor could refuse to compile the module, if it is in fixed-form. Nevertheless, people could still compile and develop their legacy codes, they just would not be allowed to use any new constructs (and default behaviors) introduced in Fortran 202Y or later without converting to free-form.

**Comment by septcolor** [comment]
@sblionel wrote:
> (...snip...) Keep in mind that every compiler implementor I have spoken with on the topic of deleted features has said that they must continue to support deleted features. Deleting features provides no benefits and actually adds complexity for implementations which have to figure out how those features continue to work with the rest of the standard.

> My recommendation would be to "save your strength" and advocate for new features that will support the programming paradigms you'd like to use.

Here I mostly agree with Steve's opinions above (regarding fixed form)... Indeed,  I am skeptical that the (hypothetical) deletion of fixed form will actually contribute to giving any fresh impression for new users or encouraging new projects in Fortran. This is because many users already know free form and have been using it (unless fixed form is enforced by existing projects). Nevertheless, there is a continuing decline in the number of new projects  (according to #106 and also from my experience...), suggesting that key problems are not fixed vs free form.

On technical sides, I also think it is rather easy to make new syntaxes to fit with fixed form (e.g., #102). For enumerators (#11), simply switching the keywords ENUMERATOR and ENUM in may make it  less verbose, even if we cannot omit ENUM inside the construct (and there also seem to be more concise syntax possible...).

> I do not agree that fixed-form source is why (allegedly) young programmers are not learning Fortran. I also disagree with that basic premise. While Fortran is not often taught in the US, in favor of sparkly languages not common in production environments, it is widely taught in Europe and Asia. This isn't the standard's fault.

Here, my experience is very different from Steve's assumption above. For example, in our (combined) research group, the ratio of students who use Fortran has significantly dropped in the last 10 years because new faculty staffs are using C++ (and so their students as well). Another factor that accelerates this trend is the retirement of old professors who "enforced" students to use Fortran (apart from it is good or not).

I am also not sure about the popularity of Fortran in Asia now; for example, there seems to be a growing community of Julia in China (with a dedicated Discourse site in Chinese https://discourse.juliacn.com/). In Japan, there are few books of Fortran sold in book stores (which can be noticed very easily). As for UK, I remember having read a web page showing that the fraction of Fortran programs used in ARCHER (a UK supercomputer) is now about 50 % or so (? I will search for the original article if necessary), while Fortran dominated such supercomputers in ~2000 (IIRC). Even that 50% is very likely dominated by well-established program suites (e.g., VASP for materials science), and the ratio of new programs is unknown (at least I did not find any new program names in the application list shown in the web page). So, my experience is "not at all optimistic", unlike your observation. I also think such a situation of course relates to the current standard, at least partly (again IMHO).

**Comment by septcolor** [comment]
@aradi wrote:
> I'd like to note that a language version keyword as discussed in #83 could also allow for giving up compatibility with fixed-form source. Whenever the programmer specifies a language version newer than a given date (e.g. 202Y), the processor could refuse to compile the module, if it is in fixed-form. Nevertheless, people could still compile and develop their legacy codes, they just would not be allowed to use any new constructs (and default behaviors) introduced in Fortran 202Y or later without converting to free-form.

I think the above approach is very appealing, but I am wondering whether a parser (of Fortran sources) may need to assume fixed or free form before starting to parse them anyway by other means? (otherwise, it may be impossible to do parsing itself...?)

**Comment by sblionel** [comment]
The typical way of handling this is a directive that looks like a whole-line comment. You're already familiar with these (!DIR$, !OMP$, etc.), as well as shell scripts. Many compilers already support `!DIR$ FREEFORM` or similar.

I don't recall if J3 has ever taken up the idea of directives in Fortran source, rather than preprocessing.

**Comment by certik** [comment]
@sblionel when it comes to adding new features to Fortran, do we still need to consider how they will work with fixed-form? Is it acceptable to simply say that for new features, such as #11, we do not need to consider fixed-form? I see lots of positives of that, but I don't know if there is any downside.

**Comment by sblionel** [comment]
Yes, we still need to consider fixed-form. It is still part of the language. And despite my entreaties to the contrary, people still write new code in fixed-form.

**Comment by certik** [comment]
> Yes, we still need to consider fixed-form. It is still part of the language. And despite my entreaties to the contrary, people still write new code in fixed-form.

What would be the reasons why people have to write fixed-form? Is it just a lack of a tool that would reliably do the conversion, or is there another reason?

**Comment by sblionel** [comment]
"have to"? No. But it's what they're used to and that's what they do.

**Comment by gronki** [comment]
That's what I mentioned in some other posts. We all know they do. But they
shouldn't. How do we force them to stop? I think not having nice features
would be a good motivation.

In Linux world, people are unhappy with changes all the time, but then the
new solution is forced upon users and everyone just adjusts. I think
standard is not enforcing enough to discourage bad coding such as fixed
form.

czw., 12 gru 2019, 16:51 użytkownik Steve Lionel <notifications@github.com>
napisał:

**Comment by certik** [comment]
When I discussed this with quite a few people, it seems that a reasonable compromise is to support fixed-form for current codes, but not adding new features to it, e.g., if you want to use enumerators, you would first need to convert your fixed-form to free form. So that we do not need to discuss fixed-form restrictions when designing a new feature.

**Comment by certik** [comment]
@FortranFan wrote:

> May I please see some consistent adherence and application of this?

I just created a separate issue #109 to discuss how to communicate, please let's move that discussion there, and only discuss fixed-form here in this issue.

**Comment by FortranFan** [comment]
@certik wrote:

> When I discussed this with quite a few people, it seems that a reasonable compromise is to support fixed-form for current codes, but not adding new features to it, e.g., if you want to use enumerators, you would first need to convert your fixed-form to free form. So that we do not need to discuss fixed-form restrictions when designing a new feature.

Ergo the proposal in the original post.

Within the parameters of the current system toward the Fortran standard, the compromise of "support fixed-form for current codes, but not adding new features to it" only appears feasible if fixed-form is marked as DELETED in the official standard.

Deleting fixed-form from the standard will not preclude any existing or new compiler from supporting fixed-form: one can look at the support of already deleted features in most of the current compilers.  But the further development of the standard is freed up from the confines of fixed-form.

As indicated in https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-565061961 with the committee approach of which I've been acutely aware, my worry all along has been even the features toward Fortran 202Y - say GENERICS and perhaps exception handling or one or two others if they get considered - would all need to be compatible in their design with fixed-form source.  My experience remains the need to remain compatible with fixed-form has adversely impacts VERBOSITY in Fortran.

Unless the compromise indicated by @certik can be pursued with some urgency, there remains the risk of suboptimal designs of new features in Fortran due to support for fixed-form source and that is not at all a healthy and encouraging prospect for developers like @jacobwilliams who are among the few persevering with Fortran.

**Comment by klausler** [comment]
> That's what I mentioned in some other posts. We all know they do. But they shouldn't. How do we force them to stop? I think not having nice features would be a good motivation. In Linux world, people are unhappy with changes all the time, but then the new solution is forced upon users and everyone just adjusts. I think standard is not enforcing enough to discourage bad coding such as fixed form.

Why in the world would anybody want to *force* fixed-form users to stop using fixed-form?

**Comment by gronki** [comment]
For the reasons clearly listed above that I am not going to repeat. :)

czw., 12 gru 2019 o 18:08 Peter Klausler <notifications@github.com>
napisał(a):

**Comment by klausler** [comment]
> For the reasons clearly listed above that I am not going to repeat. :) 

The only serious point that was raised was that fixed-form makes it harder to add a feature to the language that would change parsing of keyword statements like `END` to be context-dependent.  Such bad ideas *should* be hard to add to the language, so that's a reason to *retain* fixed-form.

**Comment by certik** [comment]
@klausler let's discuss the actual argument in #11 more with regards to fixed-form. One proposed syntax is:
```fortran
enum some_type
    enumerator a, b, c, d, end, e, endenum, f
end enum
```
and the reason why we cannot just have
```fortran
enum some_type
    a, b, c, d, end, e, endenum, f
end enum
```
is that `end` and `endenum` can be confused with `end enum` when parsing (especially if the optional space is omitted). I think that's probably true, but I think the modern standard could be modified to require `end if`, `end function`, `end enum` to be always written with a space in between. And in that case, I don't know if there is still an ambiguity in free-form (more ambiguity than there already is with regards to `end` -- it is already context dependent)

**Comment by klausler** [comment]
```
ENUM SOME_TYPE
  CONTINUE
  STOP
  END
  CONTAINS
  PRIVATE
  SAVE
END ENUM
```

Statement classification is already difficult and bug-prone in Fortran.  This feature makes it worse -- a bare `END` statement may no longer be an `END` statement, &c.

**Comment by certik** [comment]
@klausler I think you are right. This is the main argument, and that has nothing to do with fixed-form. 

I propose that fixed-form arguments should not even enter the discussion, and should not be mentioned in the proposal for #11. Rather, the argument that you just made about the `end` ambiguity is the main argument.

**Comment by FortranFan** [comment]
@certik wrote:

> @FortranFan's main point is valid, that fixed form was used as an argument why something cannot be done in 19/19-249.txt. If fixed form didn't have to be considered, perhaps the `ENUMERATOR` keyword would still be necessary for other reasons, as @klausler indicated, but perhaps not. However let's discuss that at #11.
> 
> @FortranFan please try hard to stick on topic and just discuss technical arguments in issues, otherwise every issue becomes buried and people stop participating here. If you want to discuss your wider points, just open a new issue and link it. But for already open issues that somebody opened to discuss a particular topic, please try not to derail the discussion.

In #92 with https://github.com/j3-fortran/fortran_proposals/issues/92#issuecomment-569315654, @klausler wrote:

> Attributes before the `FUNCTION` and `SUBROUTINE` keyword are how things are done today, like it or not. It was a bad design and it restricted at least one syntactic possibility (one can write `CHARACTER*7` or `CHARACTER(LEN=NAMEDCONSTANT)` but not `CHARACTER*NAMEDCONSTANT`, and that's due at least in part to the ambiguity that would introduce in `CHARACTER*IMPUREFUNCTIONFOO()`). But that ship has sailed; adding a redundant `ATTRIBUTES(...)` syntax would not enable any new capability.

```CHARACTER*NAMEDCONSTANT``` does not conform to the standard but ```CHARACTER(NAMEDCONSTANT)``` does.

And a function declaration such as ```CHARACTER(NAMEDCONSTANT)IMPUREELEMENTALFUNCTIONFOO()``` is standard-conforming in **fixed-form source** (assuming other requirements are satisfied) but obviously not in free-form source.  Regardless I find it irrelevant to the discussion in #92 on FUNCTION or SUBROUTINE declarations.

Per section B.3.7 in 18-007r1 document toward Fortran 2018, fixed-form is obsolescent.

Bringing up a fixed-form code snippet to influence direction in any manner in a discussion to a new proposal (whatever "tier" it may be in, a convenience to a few mere mortal but persevering Fortranners or otherwise) which then also presents a risk to "derail the discussion" is exactly what I was trying to warn in my above https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-564713380.  It happens often enough, it's harmful in that it too stops people from "participating" in Fortran.  OP's proposal in this thread to "Delete fixed-form source format" remains a good suggestion for Fortran.

**Comment by sblionel** [comment]
The existence of fixed form can affect syntax chosen, but it has never, to my knowledge, actually blocked a feature from being developed. (19-249 is not an example of such.)

I'll also mention: "The program unit END statement shall not be continued. A statement whose initial line appears to be a program unit END statement shall not be continued." (6.3.3.5) 

Fixed form isn't why the `ENUMERATOR` keyword exists, as @klausler points out. Fortran doesn't have reserved words and isn't going to start adding them.

I would be cautiously agreeable to the notion that new features be available in free-form only, where it made sense to do so. As I've mentioned before, there's a significant cost to deleting features from the language.

**Comment by jacobwilliams** [comment]
I would think the easiest way to ensure that new features only apply to free-form is to just delete fixed-form from the standard. Then you never have to worry about it again (let the vendors who want to worry about maintaining it for people who need to run unmodified 50 year old code...but the standard would be free to move forward unencumbered).

**Comment by klausler** [comment]
> I would think the easiest way to ensure that new features only apply to free-form is to just delete fixed-form from the standard. Then you never have to worry about it again (let the vendors who want to worry about maintaining it for people who need to run unmodified 50 year old code...but the standard would be free to move forward unencumbered).

Suppose for the sake of discussion that fixed source form were to be removed from the standard.  Which Fortran compilers do you think would remove support for fixed source form?

**Comment by everythingfunctional** [comment]
> > I would think the easiest way to ensure that new features only apply to free-form is to just delete fixed-form from the standard. Then you never have to worry about it again (let the vendors who want to worry about maintaining it for people who need to run unmodified 50 year old code...but the standard would be free to move forward unencumbered).
> 
> Suppose for the sake of discussion that fixed source form were to be removed from the standard. Which Fortran compilers do you think would remove support for fixed source form?

My guess is the same number that have removed support for non-standards extensions (effectively zero?). But a guy can hope.

**Comment by FortranFan** [comment]
@sblionel wrote:

> .. there's a significant cost to deleting features from the language.

I request any argument in terms of "cost" be treated as firmly off-topic in this repository.

**No reliable and accurate data** are available on any matter related to costs in Fortran.  No honest or trustworthy discussion is then possible.

Moreover, any discussion regarding "cost" is meaningless without a full and fair consideration of the benefits.

When it comes to benefits though, I find the feedback on the value proposition and the benefit statements offered by many a persevering, "mere mortal" Fortranner either gets ignored or is rejected outright.

Under the circumstances, no "good faith" discussion can take place once cost and only cost makes it way into the discussion.

Yet if the cost argument is allowed to persist, then a honest position will be to note:

* features delayed are features denied, period.  The resultant "opportunity costs" to Fortran and its well-being are then immeasurably high.  Entire code-bases move permanently away from Fortran as a result

* features ***HALF-BAKED*** (e.g., several aspects introduced in Fortran 2003) are features ignored or misused.  The resultant "opportunity costs" to Fortran and its well-being are again immeasurably high.

* features too ***VERBOSE*** are features abhorred by a large section of the computing community.  The costs to Fortran's well-being again are immense.

With Fortran, all 3 of these points - delays, half-baked aspects, and verbosity - have been impacted to some extent or other by vendors' notions of backward compatibility, the continuing support of fixed-form source very much falls under this.

There is considerable evidence an inherently dishonest and unequal treatment of backward compatibility remains in Fortran, one just has to look at Annex B of the Fortran standard.

The aspect of the **obsolescent-since-1997 fixed-form source** facility is part and parcel of this biased treatment of backward compatibility in the language.

The cost argument is simply untenable, more than anything else it's one that always threatens to "derail a discussion".

**Comment by certik** [comment]
@FortranFan I agree with you that we should evaluate both the "cost" as well as the "opportunity cost". (Although I think it's very much on-topic here to discuss these costs.)

So I think @sblionel is correct that "there's a significant cost to deleting features from the language". But I also believe there is a significant opportunity cost of not deleting features from the language, and we have to weigh both.

**Comment by klausler** [comment]
At what point on a language change's path from concept to formal description to standardization to implementation should cost become relevant?

**Comment by Meriipu** [comment]
> (...) 
> and the reason why we cannot just have
> 
> ```fortran
> enum some_type
>     a, b, c, d, end, e, endenum, f
> end enum
> ```
> 
> is that `end` and `endenum` can be confused with `end enum` when parsing (especially if the optional space is omitted). I think that's probably true, **but I think the modern standard could be modified to require `end if`, `end function`, `end enum` to be always written with a space in between.**

unless I misunderstand, this would break even more code (including conforming free-form code) than some of the more conservative suggestions? And perhaps be excessively normative about code style.

**Comment by sblionel** [comment]
@Meriipu, no, that's not correct. The only restriction is that a source line cannot look like an END statement (see @klausler 's [post above](https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-565115131)). The syntax you show is acceptable in fixed and free form (ignoring the `some_type` which is not valid - F202X adds a new form of enumeration types with names, but the syntax is a bit different.)

**Comment by Meriipu** [comment]
I was referring to the suggestion at the end of the quoted text (emphasized in bold) from the comment https://github.com/j3-fortran/fortran_proposals/issues/108#issuecomment-565111834 which I took to understand to start enforce the optional space.

**Comment by sblionel** [comment]
Oops - failed to read carefully. Sorry.  

I'm still of the opinion that there's no good reason to delete fixed form, nor introduce new features valid only in free form. Yes, one can construct syntax that doesn't work, but so what? Other syntax, consistent with the language, is possible, as repeatedly demonstrated over the years.
